Cirrus: ruby3.0 build error on macOS

MacOS builds are failing on CirrusCI because ruby@2.7 was updated to ruby@3.0 and the 2.7 tap is not available anymore on homebrew (see [cirrus](https://cirrus-ci.com/task/4888409065914368)).
I tested ruby@3.0 locally and it seems that ruby 3.0 causes some compatibility problems which causes the build to fail.

```
/usr/local/Cellar/ruby/3.0.0/include/ruby-3.0.0/ruby/backward.h:35:1: error: '__declspec' attributes are not enabled; use '-fdeclspec' or '-fms-extensions'
      to enable support for __declspec attributes
DECLARE_DEPRECATED_FEATURE(2.2, rb_frame_pop);
^
/usr/local/Cellar/ruby/3.0.0/include/ruby-3.0.0/ruby/backward.h:16:5: note: expanded from macro 'DECLARE_DEPRECATED_FEATURE'
    NORETURN(ERRORFUNC(("deprecated since "#ver), DEPRECATED(void func(void))))
    ^
/usr/local/Cellar/ruby/3.0.0/include/ruby-3.0.0/ruby/backward/2/attributes.h:152:21: note: expanded from macro 'NORETURN'
#define NORETURN(x) RBIMPL_ATTR_NORETURN() x
                    ^
/usr/local/Cellar/ruby/3.0.0/include/ruby-3.0.0/ruby/internal/attr/noreturn.h:29:33: note: expanded from macro 'RBIMPL_ATTR_NORETURN'
# define RBIMPL_ATTR_NORETURN() __declspec(noreturn)
                                ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
make[2]: *** [src/plugins/ruby/CMakeFiles/elektra-ruby-objects.dir/ruby.cpp.o] Error 1
make[1]: *** [src/plugins/ruby/CMakeFiles/elektra-ruby-objects.dir/all] Error 2
make: *** [all] Error 2
make: INTERNAL: Exiting with 11 jobserver tokens available; should be 10!
```

--- a/.cirrus.yml
+++ b/.cirrus.yml
@@ -166,8 +166,8 @@ mac_task:
     - | # Install Python packages
       pip3 install cmake-format[yaml]==0.6.3
     - | # Install Ruby (We use the Homebrew version of Ruby to install gems, since compiling `ronn` fails with the macOs version of Ruby.)
-      brew install ruby@2.7
-      export PATH="/usr/local/opt/ruby@2.7/bin:/usr/local/lib/ruby/gems/2.7.0/bin:$PATH"
+      brew install ruby
+      export PATH="/usr/local/opt/ruby/bin:/usr/local/lib/ruby/gems/3.0.0/bin:$PATH"
     - | # Install Ruby gems
       sudo gem install ronn test-unit --no-document
     - | # Install JavaScript packages

--- a/.github/workflows/macOS_gcc10.yml
+++ b/.github/workflows/macOS_gcc10.yml
@@ -9,6 +9,7 @@ env:
   # Unfortunately the tests for the Xerces plugin fail: https://travis-ci.org/ElektraInitiative/libelektra/jobs/483331657#L3740
   # The curlget tests fail: https://github.com/ElektraInitiative/libelektra/issues/3382
   PLUGINS: 'ALL;-xerces;-curlget'
+  BINDINGS: 'ALL;-rust'
   # Skip homebrew cleanup to avoid issues with removal of packages
   HOMEBREW_NO_INSTALL_CLEANUP: 1
   CC: gcc-10

--- a/.github/workflows/macOS_gcc10.yml
+++ b/.github/workflows/macOS_gcc10.yml
@@ -1,0 +1,103 @@
+name: macOS GCC
+
+# Triggers the workflow on push or pull request events
+on: [push, pull_request]
+
+env:
+  # Customize the CMake build type here (Release, Debug, RelWithDebInfo, etc.)
+  BUILD_TYPE: RelWithDebInfo
+  # Unfortunately the tests for the Xerces plugin fail: https://travis-ci.org/ElektraInitiative/libelektra/jobs/483331657#L3740
+  # The curlget tests fail: https://github.com/ElektraInitiative/libelektra/issues/3382
+  PLUGINS: 'ALL;-xerces;-curlget'
+  # Skip homebrew cleanup to avoid issues with removal of packages
+  HOMEBREW_NO_INSTALL_CLEANUP: 1
+
+jobs:
+  build:
+    # The CMake configure and build commands are platform agnostic and should work equally
+    # well on Windows or Mac.  You can convert this to a matrix build if you need
+    # cross-platform coverage.
+    # See: https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions/managing-complex-workflows#using-a-build-matrix
+    runs-on: macos-latest
+
+    steps:
+    - uses: actions/checkout@v2
+
+    - name: Install Dependencies
+      run: |
+        brew install augeas
+        brew install bison
+        brew install dbus
+        brew install discount
+        brew install flex
+        brew install gpgme
+        brew install libev
+        brew install libgcrypt
+        brew install libgit2
+        brew install libuv
+        brew install lua
+        brew install ninja
+        brew install qt
+        brew install swig
+        brew install yajl
+        brew install zeromq
+
+    - name: Setup Build Environment
+      run: |
+        gem install test-unit --no-document
+        export CC=gcc-10
+        export CXX=g++-10
+        pip2 install cheetah # Required by kdb-gen
+        brew services start dbus
+        cmake -E make_directory ${{runner.workspace}}/build
+
+    - name: Configure CMake
+      # Use a bash shell so we can use the same syntax for environment variable
+      # access regardless of the host operating system
+      shell: bash
+      working-directory: ${{runner.workspace}}/build
+      # Note the current convention is to use the -S and -B options here to specify source 
+      # and build directories, but this is only available with CMake 3.13 and higher.  
+      # The CMake binaries on the Github Actions machines are (as of this writing) 3.12
+      run: |
+        SYSTEM_DIR="$PWD/kdbsystem"
+        CMAKE_OPT+=(
+          -GNinja
+          -DPLUGINS="${PLUGINS:-ALL}"
+          -DBINDINGS="${BINDINGS:-ALL}"
+          -DTOOLS="${TOOLS:-ALL}"
+          -DBUILD_FULL="${BUILD_FULL:-OFF}"
+          -DBUILD_SHARED="${BUILD_SHARED:-ON}"
+          -DENABLE_ASAN="${ENABLE_ASAN:-OFF}"
+          -DCOMMON_FLAGS="${COMMON_FLAGS--Werror}"
+          -DENABLE_DEBUG=ON
+          -DCMAKE_EXPORT_COMPILE_COMMANDS=ON
+          -DKDB_DB_SYSTEM="$SYSTEM_DIR"
+          -DKDB_DEFAULT_STORAGE="${KDB_DEFAULT_STORAGE:-dump}"
+          -DKDB_DB_FILE="${KDB_DB_FILE:-default.ecf}"
+          -DKDB_DB_INIT="${KDB_DB_INIT:-elektra.ecf}"
+          $TRAVIS_BUILD_DIR
+        )
+        printf '—— CMake Config ——\n'
+        for option in "${CMAKE_OPT[@]}"; do printf '%s\n' "$option"; done
+        cmake $GITHUB_WORKSPACE -DCMAKE_BUILD_TYPE=$BUILD_TYPE ${CMAKE_OPT[@]}
+
+    - name: Build
+      working-directory: ${{runner.workspace}}/build
+      shell: bash
+      # Execute the build.  You can specify a specific target with "--target <NAME>"
+      run: ninja
+
+    - name: Install
+      working-directory: ${{runner.workspace}}/build
+      shell: bash
+      run: output="$(ninja install 2>&1)" || printf '%s' "$output"
+
+    - name: Test
+      working-directory: ${{runner.workspace}}/build
+      shell: bash
+      # Execute tests defined by the CMake configuration.  
+      # See https://cmake.org/cmake/help/latest/manual/ctest.1.html for more detail
+      run: |
+        ninja run_all
+        kdb run_all

--- a/.github/workflows/macOS_gcc10.yml
+++ b/.github/workflows/macOS_gcc10.yml
@@ -49,7 +49,9 @@ jobs:
       run: |
         gem install test-unit --no-document
         pip2 install cheetah # Required by kdb-gen
-        brew services start dbus
+        brew tap homebrew/services
+        brew update # Work around for [Homebrew Services issue 206](https://github.com/Homebrew/homebrew-services/issues/206)
+        brew services restart dbus
         cmake -E make_directory ${{runner.workspace}}/build
 
     - name: Configure CMake

--- a/.github/workflows/macOS_gcc10.yml
+++ b/.github/workflows/macOS_gcc10.yml
@@ -11,6 +11,8 @@ env:
   PLUGINS: 'ALL;-xerces;-curlget'
   # Skip homebrew cleanup to avoid issues with removal of packages
   HOMEBREW_NO_INSTALL_CLEANUP: 1
+  CC: gcc-10
+  CXX: g++-10
 
 jobs:
   build:
@@ -45,8 +47,6 @@ jobs:
     - name: Setup Build Environment
       run: |
         gem install test-unit --no-document
-        export CC=gcc-10
-        export CXX=g++-10
         pip2 install cheetah # Required by kdb-gen
         brew services start dbus
         cmake -E make_directory ${{runner.workspace}}/build

--- a/.github/workflows/macOS_gcc10.yml
+++ b/.github/workflows/macOS_gcc10.yml
@@ -61,6 +61,8 @@ jobs:
       # and build directories, but this is only available with CMake 3.13 and higher.  
       # The CMake binaries on the Github Actions machines are (as of this writing) 3.12
       run: |
+        PATH="/usr/local/opt/ruby@2.7/bin:/usr/local/lib/ruby/gems/2.7.0/bin:$PATH"
+        PATH="/usr/local/opt/python@3.9/bin:$PATH"
         SYSTEM_DIR="$PWD/kdbsystem"
         CMAKE_OPT+=(
           -GNinja

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,6 +1,5 @@
 language: cpp
 dist: bionic
-osx_image: xcode12.2
 
 #
 # Define the build matrix
@@ -151,38 +150,6 @@ matrix:
         - ninja run_all
         - kdb-full run_all
 
-    - os: osx
-      name: 🍏 GCC
-      compiler: gcc
-      addons:
-        homebrew:
-          packages:
-            - augeas
-            - bison
-            - dbus
-            - discount
-            - flex
-            - gpgme
-            - libev
-            - libgcrypt
-            - libgit2
-            - libuv
-            - lua
-            - ninja
-            - openssl
-            - python@3.9
-            - qt
-            - swig
-            - yajl
-            - zeromq
-          update: true
-      env:
-        # Unfortunately the tests for the Xerces plugin fail: https://travis-ci.org/ElektraInitiative/libelektra/jobs/483331657#L3740
-        # The curlget tests fail: https://github.com/ElektraInitiative/libelektra/issues/3382
-        - PLUGINS='ALL;-xerces;-curlget'
-        # Skip homebrew cleanup to avoid issues with removal of packages
-        - HOMEBREW_NO_INSTALL_CLEANUP=1
-
     - os: linux
       name: 🐧 GCC
       compiler: gcc
@@ -211,20 +178,6 @@ matrix:
         - CXX_COMPILER=clang++-7
 
 before_install:
-  - |
-    if [ "$TRAVIS_OS_NAME" = 'osx' ]; then
-      rvm install 2.7.1
-      rvm use 2.7.1
-      gem install test-unit --no-document
-      if [ "$CC" = 'gcc' ]; then
-        brew upgrade gcc
-        export CC=gcc-10
-        export CXX=g++-10
-      fi
-      pip2 install cheetah # Required by kdb-gen
-      # start D-Bus session bus
-      brew services start dbus
-    fi
   - |
     if [ "$TRAVIS_OS_NAME" = 'linux' ]; then
       [ -n "$CC_COMPILER" ] && export CC="$CC_COMPILER"

--- a/doc/news/_preparation_next_release.md
+++ b/doc/news/_preparation_next_release.md
@@ -193,7 +193,7 @@ you up to date with the multi-language support provided by Elektra.
 
 ### Travis
 
-- <<TODO>>
+- Move macOS GCC 10 build job to Github Actions. _(Mihael Pranjić)_
 - <<TODO>>
 - <<TODO>>
 

--- a/doc/news/_preparation_next_release.md
+++ b/doc/news/_preparation_next_release.md
@@ -131,7 +131,9 @@ you up to date with the multi-language support provided by Elektra.
 
 - Remove ipairs support and add our own iterator to add support for Lua 5.4, since `__ipairs` was deprecated. _(Manuel Mausz)_
 
-### <<Binding2>>
+### Ruby
+
+- Enable `__declspec` attributes for Ruby 3.0. _(Mihael Pranjić)_
 
 ### <<Binding3>>
 
@@ -181,7 +183,7 @@ you up to date with the multi-language support provided by Elektra.
 ### Cirrus
 
 - Upgrade Cirrus Fedora docker image to Fedora 33. _(Mihael Pranjić)_
-- <<TODO>>
+- Upgrade to Ruby 3.0 for macOS builds. _(Mihael Pranjić)_
 - <<TODO>>
 
 ### Jenkins

--- a/scripts/kdb/mount-list-all-files
+++ b/scripts/kdb/mount-list-all-files
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/usr/bin/env sh
 #
 # @tags mount
 
@@ -7,7 +7,7 @@ if [ -z "$KDB" ]; then
 fi
 
 MOUNTPOINTS=$(kdb complete system:/elektra/mountpoints/ --max-depth=1)
-MOUNTPOINTS=$(echo "$MOUNTPOINTS" | sed 's:system:/elektra/mountpoints/::' | sed 's:\\/:/:g')
+MOUNTPOINTS=$(echo "$MOUNTPOINTS" | sed 's#system:/elektra/mountpoints/##' | sed 's:\\/:/:g')
 FILES=
 OLD_IFS="$IFS"
 IFS="$(printf '\n+')"

--- a/src/bindings/swig/ruby/CMakeLists.txt
+++ b/src/bindings/swig/ruby/CMakeLists.txt
@@ -69,6 +69,7 @@ else ()
 	# disable certain compiler warnings for SWIG generated files
 	set (SWIG_COMPILE_FLAGS "${SWIG_COMPILE_FLAGS} -Wno-unused-parameter")
 	set (SWIG_COMPILE_FLAGS "${SWIG_COMPILE_FLAGS} -Wno-sign-compare")
+	set (SWIG_COMPILE_FLAGS "${SWIG_COMPILE_FLAGS} -fdeclspec")
 	if (CMAKE_CXX_COMPILER_ID MATCHES "GNU")
 		set (SWIG_COMPILE_FLAGS "${SWIG_COMPILE_FLAGS} -Wno-unused-but-set-variable")
 		set (SWIG_COMPILE_FLAGS "${SWIG_COMPILE_FLAGS} -Wno-catch-value")

--- a/src/bindings/swig/ruby/CMakeLists.txt
+++ b/src/bindings/swig/ruby/CMakeLists.txt
@@ -69,7 +69,6 @@ else ()
 	# disable certain compiler warnings for SWIG generated files
 	set (SWIG_COMPILE_FLAGS "${SWIG_COMPILE_FLAGS} -Wno-unused-parameter")
 	set (SWIG_COMPILE_FLAGS "${SWIG_COMPILE_FLAGS} -Wno-sign-compare")
-	set (SWIG_COMPILE_FLAGS "${SWIG_COMPILE_FLAGS} -fdeclspec")
 	if (CMAKE_CXX_COMPILER_ID MATCHES "GNU")
 		set (SWIG_COMPILE_FLAGS "${SWIG_COMPILE_FLAGS} -Wno-unused-but-set-variable")
 		set (SWIG_COMPILE_FLAGS "${SWIG_COMPILE_FLAGS} -Wno-catch-value")
@@ -92,6 +91,7 @@ else ()
 		set (SWIG_COMPILE_FLAGS "${SWIG_COMPILE_FLAGS} -Wno-deprecated-register")
 		set (SWIG_COMPILE_FLAGS "${SWIG_COMPILE_FLAGS} -Wno-gnu-zero-variadic-macro-arguments")
 		set (SWIG_COMPILE_FLAGS "${SWIG_COMPILE_FLAGS} -Wno-zero-length-array")
+		set (SWIG_COMPILE_FLAGS "${SWIG_COMPILE_FLAGS} -fdeclspec")
 	endif ()
 
 	# add 'kdb' module

--- a/src/plugins/ruby/CMakeLists.txt
+++ b/src/plugins/ruby/CMakeLists.txt
@@ -43,7 +43,7 @@ if (DEPENDENCY_PHASE)
 		add_custom_command (OUTPUT runtime.h COMMAND ${SWIG_EXECUTABLE} -c++ -ruby -external-runtime runtime.h)
 
 		# we call this SWIG_COMPILE_FLAGS because we have the same variable in our swig bindings
-		set (SWIG_COMPILE_FLAGS "-Wno-shadow -Wno-old-style-cast -Wno-unused-variable -Wno-unused-parameter")
+		set (SWIG_COMPILE_FLAGS "-Wno-shadow -Wno-old-style-cast -Wno-unused-variable -Wno-unused-parameter -fdeclspec")
 
 		# Disable warnings produced by generated code
 		if (CMAKE_CXX_COMPILER_ID MATCHES "Clang")

--- a/src/plugins/ruby/CMakeLists.txt
+++ b/src/plugins/ruby/CMakeLists.txt
@@ -43,7 +43,7 @@ if (DEPENDENCY_PHASE)
 		add_custom_command (OUTPUT runtime.h COMMAND ${SWIG_EXECUTABLE} -c++ -ruby -external-runtime runtime.h)
 
 		# we call this SWIG_COMPILE_FLAGS because we have the same variable in our swig bindings
-		set (SWIG_COMPILE_FLAGS "-Wno-shadow -Wno-old-style-cast -Wno-unused-variable -Wno-unused-parameter -fdeclspec")
+		set (SWIG_COMPILE_FLAGS "-Wno-shadow -Wno-old-style-cast -Wno-unused-variable -Wno-unused-parameter")
 
 		# Disable warnings produced by generated code
 		if (CMAKE_CXX_COMPILER_ID MATCHES "Clang")
@@ -51,6 +51,7 @@ if (DEPENDENCY_PHASE)
 			set (SWIG_COMPILE_FLAGS "${SWIG_COMPILE_FLAGS} -Wno-macro-redefined")
 			set (SWIG_COMPILE_FLAGS "${SWIG_COMPILE_FLAGS} -Wno-reserved-user-defined-literal")
 			set (SWIG_COMPILE_FLAGS "${SWIG_COMPILE_FLAGS} -Wno-gnu-zero-variadic-macro-arguments")
+			set (SWIG_COMPILE_FLAGS "${SWIG_COMPILE_FLAGS} -fdeclspec")
 		elseif (CMAKE_CXX_COMPILER_ID MATCHES "GNU")
 			set (SWIG_COMPILE_FLAGS "${SWIG_COMPILE_FLAGS} -Wno-literal-suffix")
 			set (SWIG_COMPILE_FLAGS "${SWIG_COMPILE_FLAGS} -Wno-attributes")

--- a/src/plugins/zeromqrecv/testmod_zeromqrecv.c
+++ b/src/plugins/zeromqrecv/testmod_zeromqrecv.c
@@ -48,7 +48,7 @@ static void * createTestSocket (void)
 	usleep (TIME_HOLDOFF);
 
 	void * pubSocket = zmq_socket (context, ZMQ_PUB);
-	int result = zmq_bind (pubSocket, "tcp://*:6001");
+	int result = zmq_bind (pubSocket, "tcp://127.0.0.1:6001");
 	if (result != 0)
 	{
 		yield_error ("zmq_bind failed");

--- a/tests/shell/check_oclint.sh
+++ b/tests/shell/check_oclint.sh
@@ -15,7 +15,7 @@ test -f "@PROJECT_BINARY_DIR@/compile_commands.json" || {
 }
 
 cd "@CMAKE_SOURCE_DIR@" || exit
-oclint -p "@PROJECT_BINARY_DIR@" -no-analytics -enable-global-analysis -enable-clang-static-analyzer \
+oclint -p "@PROJECT_BINARY_DIR@" -enable-global-analysis -enable-clang-static-analyzer \
 	"src/libs/ease/array.c" \
 	"src/libs/ease/keyname.c" \
 	"src/libs/utility/text.c" \

